### PR TITLE
Driver and example for AD7928/AD7918/AD7908 adc

### PR DESCRIPTION
**Pull request by chris-durand**
Driver for AD7928 8-channel 12 bit adc. Tested in hardware. AD7918 and AD7908 should also work, but have not been tested.

### CODE REVIEW

File: examples/stm32f746g_discovery/adc_ad7928/main.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+		XPCC_LOG_INFO << "Initialize device" << xpcc::endl;
+		PT_CALL(adc.initialize());
+
+		XPCC_LOG_INFO << "Test single conversions (Ch 0-2):" << xpcc::endl;
+
+		// Initiate first conversion, result will be output during the next conversion
+		PT_CALL(adc.singleConversion(ad7928::InputChannel::Ch0));
+
+		XPCC_LOG_INFO << PT_CALL(adc.singleConversion(ad7928::InputChannel::Ch1)) << xpcc::endl;
+		XPCC_LOG_INFO << PT_CALL(adc.singleConversion(ad7928::InputChannel::Ch2)) << xpcc::endl;
```
Comment (salkinium) [inline examples/stm32f746g_discovery/adc_ad7928/main.cpp:63]: Whoa. Didn't expect that to work… The `PT_CALL` macro is black magic… 👍

File: src/xpcc/driver/adc/ad7928.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ *     * Neither the name of the Roboterclub Aachen e.V. nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ROBOTERCLUB AACHEN E.V. ''AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL ROBOTERCLUB AACHEN E.V. BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
```
Comment (salkinium) [inline src/xpcc/driver/adc/ad7928.cpp]: short license please.

File: src/xpcc/driver/adc/ad7928.hpp [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Roboterclub Aachen e.V. nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ROBOTERCLUB AACHEN E.V. ''AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL ROBOTERCLUB AACHEN E.V. BE LIABLE FOR ANY
```
Comment (salkinium) [inline src/xpcc/driver/adc/ad7928.hpp]: short license.

File: src/xpcc/driver/adc/ad7928.hpp [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef XPCC__AD7928_HPP
+#define XPCC__AD7928_HPP
```
Comment (salkinium) [inline src/xpcc/driver/adc/ad7928.hpp]: Single `_`

File: src/xpcc/driver/adc/ad7928_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Roboterclub Aachen e.V. nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ROBOTERCLUB AACHEN E.V. ''AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL ROBOTERCLUB AACHEN E.V. BE LIABLE FOR ANY
```
Comment (salkinium) [inline src/xpcc/driver/adc/ad7928_impl.hpp]: le sigh

File: src/xpcc/driver/adc/ad7928_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+	RF_END();
+}
+
+// ----------------------------------------------------------------------------
+template <typename SpiMaster, typename Cs>
+ResumableResult<ad7928::Data>
+Ad7928<SpiMaster, Cs>::singleConversion(ad7928::InputChannel channel)
+{
+	static ControlRegister_t tempConfig;
```
Comment (salkinium) [inline src/xpcc/driver/adc/ad7928_impl.hpp]: any way to use a member variable here? maybe `outBuffer`?

**Review by salkinium — CHANGES_REQUESTED** [review]
Nice!

File: examples/stm32f746g_discovery/adc_ad7928/main.cpp (hunk 1 above)
Comment (chris-durand) [reply examples/stm32f746g_discovery/adc_ad7928/main.cpp:63]: Me neither, but it works.

File: src/xpcc/driver/adc/ad7928_impl.hpp (hunk 6 above)
Comment (chris-durand) [reply src/xpcc/driver/adc/ad7928_impl.hpp]: I found a different solution that does not require an additional variable.

Comment (chris-durand) [inline src/xpcc/driver/adc/ad7928_impl.hpp]: I found a different solution that does not require an additional variable.

**Review by salkinium — APPROVED** [review]

**Comment by chris-durand** [comment]
@salkinium I'll be unavailble for the next time (on vacation, without a laptop, partly at places without mobile network coverage :scream:). If you ~~have any more to complain~~ would like to suggest improvements, please do it until tomorrow 6am 😉.

**Comment by ekiwi** [comment]
> any more to complain

_suggested improvements_

**Comment by salkinium** [comment]
I can always pretend to be you, at least until you start signing your commits. Bwahahaha.

**Comment by salkinium** [comment]
> any more to complain

Hey, it's your name in git blame. ;-P

**Comment by ekiwi** [comment]
> Hey, it's your name in git blame. ;-P

Please, let's keep this a welcoming environment.
Not everyone will understand that the two of you know each other in real life.

**Comment by salkinium** [comment]
Yes, there's a slippery slope.

**Pull request merged**